### PR TITLE
add dynamic "shape" and "list(string)" attrs to irgen

### PR DESCRIPTION
**Pull request by marcrasi**
These are the last dynamic attrs, except for function attrs.

### CODE REVIEW

File: test/TensorFlowRuntime/dynamic_attributes.swift [hunk 1]
```diff
@@ -319,4 +368,27 @@ DynamicAttributeTests.test("TFDataTypeAttribute Array<TensorDataType>") {
   check(dataset: dataset)
 }
 
+DynamicAttributeTests.test("ShapeAttribute TensorShape") {
+  let t = Tensor<Float>([5.0])
+  let result: Tensor<Float> = #tfop("EnsureShape", t, shape$shape: loadShape(),
+                                    T$dtype: Float.tensorFlowDataType)
+  expectEqual(ShapedArray<Float>([5.0]), result.array)
```
Comment (rxwei) [inline test/TensorFlowRuntime/dynamic_attributes.swift]: Why not just use tensor equality?

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 2]
```diff
@@ -1399,6 +1425,45 @@ func _TFCOpSetAttrString(_ op: CTFEOp,
...
+                             _ strings: [String]) {
+  // Collect all the strings' utf8 bytes into a single array so that we can
+  // address all the strings with a single
+  // `flattenedStringBytes.withUnsafeBufferPointer`.
+  var flattenedStringBytes: [CChar] = []
+  var lengths: [Int] = []
+  for string in strings {
+    // Don't include the null-terminator because TFE_OpSetAttrStringList uses
+    // lengths instead of null-terminators.
+    let stringBytes = string.utf8CString.dropLast()
```
Comment (mhong) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: is there a single way to assert the last byte is indeed `\0`?

**Review by mhong — APPROVED** [review]
nice!

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 2 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: It's guaranteed by the `utf8CString` API: https://developer.apple.com/documentation/swift/string/2430818-utf8cstring, so assertion isn't strictly necessary. But i'm fine either way.

Comment (rxwei) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: It's guaranteed by the `utf8CString` API: https://developer.apple.com/documentation/swift/string/2430818-utf8cstring, so assertion isn't strictly necessary. But i'm fine either way.

File: test/TensorFlowRuntime/dynamic_attributes.swift (hunk 1 above)
Comment (marcrasi) [reply test/TensorFlowRuntime/dynamic_attributes.swift]: oh cool, that seems to work

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 3]
```diff
@@ -1399,6 +1425,45 @@ func _TFCOpSetAttrString(_ op: CTFEOp,
...
+    flattenedStringBytes.append(contentsOf: stringBytes)
+    lengths.append(stringBytes.count)
+  }
+
+  // Calculate the addresses of all the strings within our single buffer, and
+  // then call TFE_OpSetAttrStringList.
+  flattenedStringBytes.withUnsafeBufferPointer { flattenedStringBytesBuffer in
+    var stringAddrs: [UnsafeRawPointer?] = []
+    var currentStringAddr = flattenedStringBytesBuffer.baseAddress.map(UnsafeRawPointer.init)
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: fixed

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Pull request merged**